### ISSUE TITLE
BDR: DBA can create two nodes with the same name

### ISSUE THREAD

**Issue by kforte90** [2ndQuadrant/bdr#45]
The DBA can create two nodes with the same name. If the DBA then realizes his mistake, tries to remove one of the duplicates (thinking that perhaps the code will remove the first name it hits).  Instead, both nodes are removed.

Here's the case:

-- BDR group has been created previously

```
SELECT bdr.bdr_group_join( 
local_node_name := 'node2',
node_external_dsn := 'port=5599 dbname=bdrdemo',
join_using_dsn := 'port=5598 dbname=bdrdemo');

SELECT bdr.bdr_group_join( 
local_node_name := 'node2',
node_external_dsn := 'port=6000 dbname=bdrdemo',
join_using_dsn := 'port=5598 dbname=bdrdemo');

-- two nodes with the same name (column 'node_name')
 SELECT * FROM  bdr.bdr_nodes;
     node_sysid      | node_timeline | node_dboid | node_status | node_name |      node_local_dsn      |    node_init_from_dsn    
---------------------+---------------+------------+-------------+-----------+--------------------------+--------------------------
 6132063344012994765 |             1 |      16385 | r           | node1     | port=5598 dbname=bdrdemo | 
 6132063386598254836 |             1 |      16385 | r           | node2     | port=5599 dbname=bdrdemo | port=5598 dbname=bdrdemo
 6132063423648504091 |             1 |      16385 | r           | node2     | port=6000 dbname=bdrdemo | port=5598 dbname=bdrdemo
(3 rows)

-- DBA tries to get rid of one of the 'node2' nodes:
SELECT bdr.bdr_part_by_node_names('{node2}');
 bdr_part_by_node_names 
------------------------
 
(1 row)

— killed both nodes ‘node2’ ('node_status' of 'k' indicates node is 'killed' or 'removed')
bdrdemo=# select * from bdr.bdr_nodes;
     node_sysid      | node_timeline | node_dboid | node_status | node_name |      node_local_dsn      |    node_init_from_dsn    
---------------------+---------------+------------+-------------+-----------+--------------------------+--------------------------
 6132063344012994765 |             1 |      16385 | r           | node1     | port=5598 dbname=bdrdemo | 
 6132063386598254836 |             1 |      16385 | k           | node2     | port=5599 dbname=bdrdemo | port=5598 dbname=bdrdemo
 6132063423648504091 |             1 |      16385 | k           | node2     | port=6000 dbname=bdrdemo | port=5598 dbname=bdrdemo
(3 rows)
```

We could either add 'node_name' to pk of bdr.bdr_nodes or add a separate unique index on 'node_name'.

**Comment by ringerc** [comment]
What should really be happening is that the node name should be passed to `bdr.internal_node_join(...)` when we create the node entry remotely. No `node_name` parameter was added to it when we introduced node names, though.

`node_name` is non-null so we can upgrade BDR, but that means this wasn't caught earlier.

I think fixing this requires:

* Adding `node_name` to `bdr.internal_node_join(...)`, creating a new function with a new signature

* Replacing the old `internal_node_join` function with a placeholder that emits an `ERROR` like:

            ERROR: Attempting to join a BDR 0.9.0 or older node to a group running 0.9.1 or newer. Update the joining node's BDR extension first.

* Add a `UNIQUE` constraint on `bdr.bdr_nodes.node_name`

* When calling `bdr.internal_node_join` during setup, pass the node name specified by the user, so we can check that it isn't already claimed on an existing node.

Note that this isn't completely race-free, because two concurrent joins with the same node name using different target nodes could still both insert rows with the same name but different (sysid,timeline,dboid).

To completely eliminate that race we have to take the global DDL lock for long enough to insert our bdr_nodes entry. By taking the DDL lock we're ensuring that replay is fully caught up (so there are no as-yet unseen new nodes) and that nobody else can insert a node.

Eventually we could take a weaker lock that forced replay and locked just bdr.bdr_nodes. A replay barrier, essentially. We're going to need those for other things to avoid the heavyness of the DDL lock eventually, but for now the DDL lock will do.
